### ISSUE TITLE
Misleading detail in the communication of this paper?

### ISSUE THREAD

**Issue by lightvector** [lan-lc/adversarial_example_of_Go#1]
(Disclosure: I am the primary author of KataGo. I posted the below to the [openreview](https://openreview.net/forum?id=yZ_JlZaOCzv), but am posting it here too, in the hopes it might be seen and be helpful to someone.)

Thanks for this interesting work! It was linked to me and brought to my attention some time ago. However, if I understand the method correctly, I would like to raise what I believe is a misleading detail in the communication of the results in this paper, which possibly the authors themselves failed to fully appreciate. While the detail does not invalidate the results, I believe it is likely that the large majority of readers of this paper will not be technically familiar enough to spot the issue, and therefore may come away with a conclusion that, even if it turns out in future work to be true, the evidence in this paper doesn't yet support.

Specifically, this paper gives the impression that AlphaZero agents in Go can be made to return incorrect answers by adding a small number of irrelevant perturbations to the board. However, there are two straightforward ways one could perturb the board:

* A. Apply perturbations to be present already in the stone pattern as of the start of the agent's context move history context window, leaving the recent move history unchanged, i.e. with the agent and opponent's recent behavior unchanged.

* B. Insert additional perturbation moves at the end of the history for all the different AlphaZero-trained nets, as if the perturbations were moves that the agent and the opponent themselves actually just now decided to play.

According to the paper, the authors choose B, but this choice has a much larger semantic effect on the state than choice A. Choice B often asks the neural net to generate a prediction conditional on the extremely strange counterfactual "for the last two moves, both players mutually agreed that two absurd moves were in fact the best moves". This gives a substantially different interpretation to the results than the one most readers may think.

To draw a close analogy:

* Suppose you were shown a Chess tactics puzzle from a real game and asked what move you thought was best, and you solved it correctly. 

* A. Then I made one irrelevant change to the corner of board that has no relevance to the tactic and as a result you failed it instead. I can reasonably say I "tricked" you and found a clear adversarial perturbation against you.

* B. Instead suppose I performed that same change by appending two bizarrely weird or bad moves to the game, but promising to you "this tactic is from a game where on the previous turn, two super-GM players played those two moves". Then, if you trusted me, then you might seriously doubt that the obvious tactic you see works, because if it did work, high-level players would never have played those weird moves on the previous turns. And it can't simply be one player blundering - *both* players agreed that the tactic wasn't worth playing and was not worth defending against. So you conclude it must not work, and start looking for weirder options that might make their moves make sense, and you get the puzzle wrong.

In case B, I could still say I tricked you, and it was your failure because the game theoretic value and/or correct move shouldn't depend on what the history was. That is the paper's definition of "semantically equivalent". But while clearly a failure from a game-theoretic perspective, the way in which I tricked you in case B is much less impressive than in case A! And I think most people would agree the semantic equivalence between the original and perturbed situation in B is substantially weaker than that of A. But I think most readers will not think carefully about this because the paper at best only discusses it briefly - indeed most colleagues I've personally talked to who read this paper in fact did come away with the impression that A was what was claimed instead of B, thereby ending up misinformed.

There is a section "Agents are sensitive to the ordering of actions in the trajectory" with an accompanying experiment that partly addresses this, but I'm not sure it addresses the fundamental issue. The proposed move order shuffling still leaves the perturbation moves visible in the recent history. Moreover, I'm not sure this section helps the reader understand intuitively how semantically unusual the such perturbations could be, in the same way the Chess puzzle analogy above does.

The choice of B rather than A also may influence other points in the paper. For example, the point that the attack model is important because "such examples may be explored during large-scale MCTS and end up misleading the search". If it often generates states where both players' last moves forced to be directly observed blunders, perhaps B could be especially *unlikely* to generate relevant states. In a minimax tree, states on the optimal line and where only one player blunders are relevant to proving the value of the root, but states reached after *both* players blunder are entirely irrelevant to proving the root value.

However I think the efforts of the authors to do research on this topic is great! There are many known weaknesses in seemingly-superhuman agents in Go that are under-appreciated by the research community, which even affect routine usage of these agents as game analysis tools, and likely many other weaknesses not yet known. I and other on-the-ground engineers have dealt with history dependence before - for example, KataGo already has mitigations against the above issues (which I'm disappointed the authors also did not consider to enable or test!), and Leela Chess Zero devs have also considered this issue in the past. But we have not always had the attention and personal resources to do as thorough experiments as in this paper, so having such work is valuable.

It might be the case that even with A, one could still find a high density of adversarial states - that the impression readers may currently get that a small/irrelevant state change is responsible for the problem instead of history conditioning, is correct! If true, it would have more consequence for future work, since the history dependence is already a deliberate design choice more than a bug (I'm happy to elaborate on how projects like KataGo both made and/or already tried to mitigate this choice). As it stands, the experiments in this paper don't well-distinguish which of these two factors is responsible. I hope the door is open for followup work by the authors or by any other research group, to publish experiments to distinguish these two factors and/or better communication so that more readers can be educated about the results.
